cryptofeed/connection.py

```diff
@@ -164,28 +165,61 @@ def __init__(self, address: Union[List, str], conn_id: str, delay: float = 60, s
         self.sleep = sleep
         self.delay = delay
 
-    async def read(self, header=None) -> AsyncIterable:
+    async def _read_address(self, address: str, header=None) -> str:
+        if not self.is_open:
+            LOG.error('%s: connection closed in read()', self.id)
+            raise ConnectionClosed
+        LOG.debug("%s: polling %s", self.id, address)
         while True:
-            for addr in self.address:
-                if not self.is_open:
-                    LOG.error('%s: connection closed in read()', self.id)
-                    raise ConnectionClosed
-                LOG.debug("%s: polling %s", self.id, addr)
-                async with self.conn.get(addr, headers=header, proxy=self.proxy) as response:
-                    data = await response.text()
-                    self.received += 1
-                    self.last_message = time.time()
-                    if self.raw_data_callback:
-                        await self.raw_data_callback(data, self.last_message, self.id, endpoint=addr)
-                    if response.status == 429:
-                        LOG.warning("%s: encountered a rate limit for address %s, retrying in %f seconds", self.id, addr, self.delay)
-                        await asyncio.sleep(self.delay)
-                        continue
+            async with self.conn.get(address, headers=header, proxy=self.proxy) as response:
+                data = await response.text()
+                self.received += 1
+                self.last_message = time.time()
+                if self.raw_data_callback:
+                    await self.raw_data_callback(data, self.last_message, self.id, endpoint=address)
+                if response.status != 429:
                     response.raise_for_status()
-                    yield data
-                    await asyncio.sleep(self.sleep)
+                    return data
+            LOG.warning("%s: encountered a rate limit for address %s, retrying in %f seconds", self.id, address, self.delay)
             await asyncio.sleep(self.delay)
 
+    async def read(self, header=None) -> AsyncIterable[str]:
+        while True:
+            for addr in self.address:
+                yield await self._read_address(addr, header)
+            await asyncio.sleep(self.sleep)
+
+
+class HTTPConcurrentPoll(HTTPPoll):
+    """Polls each address concurrently in it's own Task"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._queue: Optional[Queue] = None
+
+    async def _poll_address(self, address: str, header=None):
+        try:
+            while True:
+                data = await self._read_address(address, header)
+                await asyncio.gather(self._queue.put(data), asyncio.sleep(self.sleep))
```

Comment:
im not sure why you are using gather here. You have already spawned a new task to run `_poll_address`, why create more tasks to put the data on the queue and then wait? cant you just `await` the queue.put and then `await` the sleep?

Reply:
The rationale was to keep the callback time more uniform to the user specified sleep time. For example, if sleep was set to 1 second, and a request takes 0.2 seconds, each callback would have a latency of 1.2 seconds. Whereas by using gather the latency would more closely resemble specified sleep time. 

Please let me know if you're not a fan of this approach and I will change the implementation.